src/main/java/io/github/thebusybiscuit/slimefun4/core/services/MetricsService.java

```diff
@@ -0,0 +1,234 @@
+            if (!download(getLatestVersion())) {
+                plugin.getLogger().warning("Failed to start metrics as the file could not be downloaded.");
+                return;
+            }
+            newlyDownloaded = true;
+        }
+
+        try {
+            // Load the jar file into a child class loader using the SF PluginClassLoader
+            // as a parent.
+            moduleClassLoader = URLClassLoader.newInstance(new URL[] { metricFile.toURI().toURL() },
+                plugin.getClass().getClassLoader());
+            Class<?> cl = moduleClassLoader.loadClass("dev.walshy.sfmetrics.MetricsModule");
+
+            metricVersion = cl.getPackage().getImplementationVersion();
+
+            // If it has not been newly downloaded, auto-updates are on AND there's a new version
+            // then cleanup, download and start
+            if (!newlyDownloaded
+                && hasAutoUpdates()
+                && checkForUpdate(metricVersion)
+            ) {
+                info("Cleaning up and re-loading Metrics.");
+                cleanUp();
+                start();
+                return;
+            }
+
+            // Finally, we're good to start this.
+            Method start = cl.getDeclaredMethod("start");
+            String s = cl.getPackage().getImplementationVersion();
+
+            // This is required to be sync due to bStats.
+            Slimefun.runSync(() -> {
+                try {
+                    start.invoke(null);
+                    info("Metrics build " + s + " started.");
+                } catch (Exception e) {
+                    plugin.getLogger().log(Level.WARNING, "Failed to start metrics.", e);
+                }
+            });
+        } catch (Exception e) {
+            plugin.getLogger().log(Level.WARNING,
+                "Failed to load the metrics module. Maybe the jar is corrupt?", e);
+        }
+    }
+
+    /**
+     * This will close the child classloader and mark all the resources held under this no longer
+     * in use, they will be cleaned up the next GC run.
+     */
+    public void cleanUp() {
+        try {
+            if (this.moduleClassLoader != null)
+                this.moduleClassLoader.close();
+        } catch (IOException e) {
+            plugin.getLogger().log(Level.WARNING,
+                "Could not clean up module class loader. Some memory may have been leaked.");
+        }
+    }
+
+    /**
+     * Checks for a new update and compares it against the current version.
+     * If there is a new version available then this returns true.
+     *
+     * @param currentVersion The current version which is being used.
+     * @return True if there is an update available.
+     */
+    public boolean checkForUpdate(String currentVersion) {
+        if (currentVersion == null || currentVersion.equals("UNOFFICIAL")) return false;
+
+        int latest = getLatestVersion();
+        if (latest > Integer.parseInt(currentVersion)) {
```

Comment:
We have a numeric pattern in PatternUtils, would be best to use that here to avoid formatting issues.